File: Angular/product-management/src/app/services/product.service.ts
import { Injectable, signal } from '@angular/core';
import { Observable, of } from 'rxjs';
import { delay, map } from 'rxjs/operators';
import { Product } from '../models/product';

@Injectable({ providedIn: 'root' })
export class ProductService {
  private readonly productsSignal = signal<Product[]>([
    { id: 1, name: 'Laptop', category: 'Electronics', price: 65000, stock: 10 },
    { id: 2, name: 'Keyboard', category: 'Accessories', price: 1500, stock: 25 },
    { id: 3, name: 'Mouse', category: 'Accessories', price: 800, stock: 40 }
  ]);

  readonly products = this.productsSignal.asReadonly();

  getAll(): Observable<Product[]> {
    return of(this.productsSignal()).pipe(
      delay(200),
      map(products => [...products])
    );
  }

  getById(id: number): Product | undefined {
    return this.productsSignal().find(product => product.id === id);
  }

  add(product: Omit<Product, 'id'>): void {
    const nextId = Math.max(0, ...this.productsSignal().map(item => item.id)) + 1;
    this.productsSignal.update(items => [...items, { id: nextId, ...product }]);
  }

  update(updated: Product): void {
    this.productsSignal.update(items =>
      items.map(item => item.id === updated.id ? updated : item)
    );
  }

  delete(id: number): void {
    this.productsSignal.update(items => items.filter(item => item.id !== id));
  }
}

File: WebApi/ProductApi/Services/ProductService.cs
using System.Collections.Concurrent;
using ProductApi.DTOs;
using ProductApi.Models;

namespace ProductApi.Services;

public class ProductService : IProductService
{
    private readonly ConcurrentDictionary<int, Product> _products = new();
    private int _nextId = 2;

    public ProductService()
    {
        _products[1] = new Product
        {
            Id = 1,
            Name = "Laptop",
            Category = "Electronics",
            Price = 65000,
            Stock = 10
        };

        _products[2] = new Product
        {
            Id = 2,
            Name = "Keyboard",
            Category = "Accessories",
            Price = 1500,
            Stock = 25
        };
    }

    public IReadOnlyCollection<Product> GetAll() =>
        _products.Values.OrderBy(p => p.Id).ToArray();

    public Product? GetById(int id) =>
        _products.TryGetValue(id, out var product) ? product : null;

    public Product Create(ProductDto dto)
    {
        var product = new Product
        {
            Id = Interlocked.Increment(ref _nextId),
            Name = dto.Name.Trim(),
            Category = dto.Category.Trim(),
            Price = dto.Price,
            Stock = dto.Stock
        };

        _products[product.Id] = product;
        return product;
    }

    public bool Update(int id, ProductDto dto)
    {
        if (!_products.TryGetValue(id, out var product))
            return false;

        product.Name = dto.Name.Trim();
        product.Category = dto.Category.Trim();
        product.Price = dto.Price;
        product.Stock = dto.Stock;
        return true;
    }

    public bool Delete(int id) => _products.TryRemove(id, out _);
}
